src/librustc_typeck/check/mod.rs

```diff
@@ -2000,6 +2001,14 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
     /// function is unreachable, and there hasn't been another warning.
     fn warn_if_unreachable(&self, id: ast::NodeId, span: Span, kind: &str) {
         if self.diverges.get() == Diverges::Always {
+            if let Node::Expr(Expr { node: ExprKind::Match(..), .. }) = self.tcx.hir.get(id) {
+                // We don't warn about `match` on `!`, because
+                // it's often desirable to explicitly match rather
+                // than taking advantage of the implicit coercions,
+                // for clarity.
+                return;
```

Comment:
I guess the idea here is that, if the match has arms, *they* will get a warning?

e.g.,

```rust
fn main() {
    return;
    match () {
        () => ()
    }
}
```

Can we get a test like that one ☝️ ?